dimod/constrained.py

```diff
@@ -1056,6 +1056,140 @@ def to_file(self, *, spool_size: int = int(1e9)) -> tempfile.SpooledTemporaryFil
+        # forbidden chars in the LP format
+        not_allowed_chars = ['+', '-', '*', '^', ':']
+        not_first_char = ['<', '>', '=', '(', ')', '[', ']', ',']
+        not_first_char += [str(i) for i in range(10)]
+
+        new_line_every = 20
+
+        f = SpooledTemporaryFile(max_size=spool_size, mode='wt')
+
+        if self.objective.linear or self.objective.quadratic:
+            f.write("Minimize\n ")
+            obj_string = 'obj: '
+
+            i_term = 0
+            if self.objective.linear:
+                for var, bias in self.objective.linear.items():
+                    obj_string += '{} {} {} '.format(sign_dict[int(np.sign(bias))], abs(bias), var)
+                    i_term += 1
+
+                    if i_term % new_line_every == 0:
+                        obj_string += '\n '
+
+            if self.objective.quadratic:
+                obj_string += '+ [ '
+                for edge, bias in self.objective.quadratic.items():
+                    var1, var2 = edge
+                    obj_string += '{} {} {} * {} '.format(sign_dict[np.sign(bias)], abs(2 * bias), var1, var2)
+                    i_term += 1
+                    if i_term % new_line_every == 0:
+                        obj_string += '\n '
+                obj_string += ']/2 '
+
+            f.write(obj_string)
+            f.write("\n \n")
+
+        if self.constraints:
+            f.write("Subject To \n")
+
+            for label, constraint in self.constraints.items():
+
+                if label[0] in not_first_char:
+                    raise ValueError('Cannot start variable name with {}'.format(label[0]))
+
+                if any([l in not_allowed_chars for l in label]):
+                    raise ValueError('Cannot use characters {} in names'.format(not_allowed_chars))
+
+                c_string = '{}: '.format(label)
+
+                i_term = 0
+                if constraint.lhs.linear:
+                    for var, bias in constraint.lhs.linear.items():
+                        c_string += '{} {} {} '.format(sign_dict[int(np.sign(bias))], abs(bias), var)
+                        i_term += 1
+
+                        if i_term % new_line_every == 0:
+                            c_string += '\n '
+
+                if constraint.lhs.quadratic:
+                    c_string += '+ [ '
+                    for edge, bias in constraint.lhs.quadratic.items():
+                        var1, var2 = edge
+                        c_string += '{} {} {} * {} '.format(sign_dict[np.sign(bias)], abs(bias), var1, var2)
+                        i_term += 1
+                        if i_term % new_line_every == 0:
+                            c_string += '\n '
+                    c_string += '] '
+
+                rhs = constraint.rhs - constraint.lhs.offset
+                c_string += ' {} {} \n'.format(sense_dict[constraint.sense.value], rhs)
+                f.write(c_string)
+
+        vartypes = {var: self.vartype(var) for var in self.variables}
```

Comment:
I think you want to do
```python
vartypes = {repr(var): self.vartype(var) for var in self.variables}
if len(vartypes) != len(cqm.variables):
    raise ValueError("some variables do not have a unique representation")
```

test case for this:
```python
variables = [(0, 1), "(0, 1)"]
```

Reply:
the only thing is that this raises an error when I look for the variable bounds because I'm calling the upper (or lower) bound method with a string

Reply:
also,  the test case above, `variables = [(0, 1), "(0, 1)"]` does not apply to LP files because those variables contain forbidden names. But still that would give me two different representations of the tuples..

Reply:
Sure, if it's handled by the "invalid name" error then we're all good.